hipcheck/src/policy_exprs/mod.rs

```diff
@@ -56,17 +59,39 @@ impl ExprMutator for Env<'_> {
 		Ok(prim.resolve(self)?.into())
 	}
 	fn visit_function(&self, f: Function) -> Result<Expr> {
+		let mut f = f;
+		// first evaluate all the children
+		f.args = f
+			.args
+			.into_iter()
+			.map(|a| self.visit_expr(a))
+			.collect::<Result<Vec<Expr>>>()?;
 		let binding = self
 			.get(&f.ident)
 			.ok_or_else(|| Error::UnknownFunction(f.ident.deref().to_owned()))?;
-		if let Binding::Fn(op) = binding {
-			(op)(self, &f.args)
+		if let Binding::Fn(op_info) = binding {
+			// Doesn't use `execute` because currently allows Functions that haven't been changed
+			// to Lambdas
+			(op_info.op)(self, &f.args)
 		} else {
 			Err(Error::FoundVarExpectedFunc(f.ident.deref().to_owned()))
 		}
 	}
-	fn visit_lambda(&self, l: Lambda) -> Result<Expr> {
-		Ok((*l.body).clone())
+	fn visit_lambda(&self, mut l: Lambda) -> Result<Expr> {
+		// Eagerly evaluate the arguments to the lambda but not the func itself
+		let Expr::Function(f) = l.body.as_mut() else {
+			unreachable!();
```

Comment:
The same question about `unreachable!()` as above.

Reply:
Added error messsage